### PR TITLE
fix: Persist "Mask fragile user data while exporting Onyx state" switch state

### PR DESCRIPTION
**Pull request by etCoderDysto**
### Fixed Issues

$ https://github.com/Expensify/App/issues/44841
PROPOSAL: https://github.com/Expensify/App/issues/44841#issuecomment-2209209223

### Tests

1. Navigate to account settings > Troubleshoot
2. Disable 'Mask fragile user data while exporting Onyx state' switch
3. Navigate away from Troubleshoot page
4. Navigate back to settings > Troubleshoot
- [x] Verify that the switch remains disabled

1. Enable 'Mask fragile user data while exporting Onyx state' switch
2. Navigate away from Troubleshoot page
3. Navigate back to settings > Troubleshoot
- [x] Verify that the switch remains enabled
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as Test steps. 
**Note:** Switch won't be grayed out when toggling it in offline mode. The feature is only client side. 
### QA Steps

Same as Test steps. 
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/9311df5d-932a-43ca-8857-25fada4f9af6

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/c6927a34-57c2-447f-a225-39b94c6553a0

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/d72a1082-1716-4b16-9ae2-fb237a2f461c

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/8482d572-8010-4266-847f-4cec1733675a

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/8503fd4d-dd5a-46a8-9b8b-bf8b4e90c62b

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/5a15fde1-d5bd-4497-ba4b-3b6fb7631149

</details>

### CODE REVIEW

**Comment by etCoderDysto** [comment]
@ZhenjaHorbach  The switch is disabled by default. Should I make it enabled by default?

**Comment by etCoderDysto** [comment]
> > @ZhenjaHorbach The switch is disabled by default. Should I make it enabled by default?
> 
> I think yes In the previous implementation it is enabled by default

Great! I am working on it.

**Comment by etCoderDysto** [comment]
> And let's create a separate file for setShouldMaskOnyxState
> I'm not sure that App.ts is the place for this

Sure! I was also not sure if I should put it there. I put it there to implement the idea. I will create a separate file.

File: src/pages/settings/Troubleshoot/TroubleshootPage.tsx [hunk 1]
```diff
@@ -40,19 +41,27 @@ type BaseMenuItem = {
...
     const illustrationStyle = getLightbulbIllustrationStyle();
-    const [shouldMaskOnyxState, setShouldMaskOnyxState] = useState(true);
+
+    useEffect(() => {
+        // Set shouldMaskOnyxState to true when user navigates to troubleshoot page after sign in
+        if (shouldMaskOnyxState !== undefined) {
+            return;
+        }
+        setShouldMaskOnyxState(true);
+    });
```
[the remark this answers is not in the document]
Comment (etCoderDysto) [reply src/pages/settings/Troubleshoot/TroubleshootPage.tsx:64]: Yes, if we don't update `shouldMaskOnyx` state to be true. We will export unmasked Onyx state below even if the switch is enabled. 

https://github.com/Expensify/App/blob/69a202e5954a6a5b3364452050f0ba40e70f7009/src/pages/settings/Troubleshoot/TroubleshootPage.tsx#L69-L73.

if I add empty dependency to `useEffect`, `setShouldMaskOnyxState` is going to called only once when the user navigates to troubleshoot page after sign in.

[the remark this answers is not in the document]
Comment (etCoderDysto) [reply src/pages/settings/Troubleshoot/TroubleshootPage.tsx:64]: Sounds good to me. To confirm that I get what you said right. 

1. I should update the above diff to 
```javascript
if (shouldMaskOnyxState ?? shouldMaskOnyxState === undefined) {
                dataToShare = ExportOnyxState.maskFragileData(value);
            }
````
2. Remove the whole useEffect code block.

Is that right?

[the remark this answers is not in the document]
Comment (etCoderDysto) [reply src/pages/settings/Troubleshoot/TroubleshootPage.tsx:64]: Alternatively, we can update the diff to 
```javascript
if (shouldMaskOnyxState !== false) {
                dataToShare = ExportOnyxState.maskFragileData(value);
            }
```

Both seems to work.

File: src/pages/settings/Troubleshoot/TroubleshootPage.tsx [hunk 2]
```diff
@@ -150,7 +142,7 @@ function TroubleshootPage({shouldStoreLogs, shouldMaskOnyxState}: TroubleshootPa
                                 <TestToolRow title={translate('initialSettingsPage.troubleshoot.maskExportOnyxStateData')}>
                                     <Switch
                                         accessibilityLabel={translate('initialSettingsPage.troubleshoot.maskExportOnyxStateData')}
-                                        isOn={shouldMaskOnyxState ?? true}
+                                        isOn={!!shouldMaskOnyxState}
```
[the remark this answers is not in the document]
Comment (etCoderDysto) [reply src/pages/settings/Troubleshoot/TroubleshootPage.tsx:145]: `shouldMaskOnyxState` is defined as `boolean | undefined` here
https://github.com/Expensify/App/blob/27e2064ce9a6bf6a023e1cfe8ca0e9b8ce54ba65/src/pages/settings/Troubleshoot/TroubleshootPage.tsx#L44
That is why I applied coercion to it. And when I try to defie `shouldMaskOnyxState` as `boolean` in the above diff typescript throws an error.

[the remark this answers is not in the document]
Comment (etCoderDysto) [reply src/pages/settings/Troubleshoot/TroubleshootPage.tsx]: Removed !!.

**Comment by etCoderDysto** [comment]
> Test steps are a bit broken and not completely filled😅
I have updated the test steps 😄. Please, let me know if there is anything I can improve.

**Review by ZhenjaHorbach — APPROVED** [review]

**Review by hayata-suenaga — APPROVED** [review]
🟢

**Pull request merged**